app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt

```diff
@@ -0,0 +1,159 @@
+package com.github.warnastrophy.core.data.repository
+
+import android.content.Context
+import android.util.Log
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.edit
+import androidx.datastore.preferences.core.stringPreferencesKey
+import androidx.datastore.preferences.preferencesDataStore
+import com.github.warnastrophy.core.data.local.StorageException
+import com.github.warnastrophy.core.model.Contact
+import com.github.warnastrophy.core.util.CryptoUtils
+import com.google.gson.Gson
+import java.util.UUID
+import kotlin.collections.iterator
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.map
+
+const val CONTACTS_DATASTORE_NAME = "contacts_encrypted"
+
+val gson = Gson()
+
+val Context.contactDataStore: DataStore<Preferences> by
+    preferencesDataStore(name = CONTACTS_DATASTORE_NAME)
+
+class ContactsRepositoryLocal(private val dataStore: DataStore<Preferences>) : ContactsRepository {
+  fun keyFor(contact: Contact) = stringPreferencesKey(contact.id)
+
+  override fun getNewUid(): String {
+    return UUID.randomUUID().toString()
+  }
+
+  override suspend fun addContact(contact: Contact): Result<Unit> {
+    var r = Result.success(Unit)
+
+    try {
+      dataStore.edit {
+        if (it.contains(keyFor(contact))) {
+          r =
+              Result.failure(
+                  StorageException.DataStoreError(
+                      Exception("Contact ${contact.id} already exists")))
+          return@edit
+        }
+
+        val ciphered = CryptoUtils.encrypt(gson.toJson(contact))
+        it[keyFor(contact)] = ciphered
+        r = Result.success(Unit)
+      }
+    } catch (e: Exception) {
+      Log.e("ContactStorage", "Error saving contact ${contact.id}", e)
+      r = Result.failure(StorageException.DataStoreError(e))
+    }
+
+    return r
```

Comment:
This function used a mutable variable `r` declared outside the `dataStore.edit {}` block and updated it inside the block. Since `edit {}` is a suspending function that can run its lambda on a different coroutine context/dispatcher, this creates a cross-thread mutation. I think this is unsafe because there is no synchronization, which can lead to stale or inconsistent results. I suggest removing the external mutable variable and instead wrapping the `edit` block in `runCatching { ... }`. This approach is thread-safe because all state is local to the coroutine and no cross-thread mutation occurs
```suggestion
    val result = runCatching {
        dataStore.edit { prefs ->
            if (prefs.contains(keyFor(contact))) {
                throw StorageException.DataStoreError(Exception("Contact ${contact.id} already exists"))
            }
            val ciphered = withContext(Dispatchers.IO) { CryptoUtils.encrypt(gson.toJson(contact)) }
            prefs[keyFor(contact)] = ciphered
        }
    }.mapFailure {
        if (it is StorageException) it else StorageException.DataStoreError(it)
    }
    return result
```